src/lib/random_oracle_input/random_oracle_input.ml

```diff
@@ -16,38 +16,228 @@ let bitstring x = {field_elements= [||]; bitstrings= [|x|]}
-  let packed_bits =
-    let packed_field_elements, remaining_bits, remaining_length =
-      Array.fold bitstrings ~init:([], [], 0)
-        ~f:(fun (acc, bits, n) bitstring ->
-          let n = n + List.length bitstring in
-          let bits = bits @ bitstring in
-          let acc, bits, n = pack_full_fields acc bits n in
-          (acc, bits, n) )
-    in
-    if remaining_length = 0 then packed_field_elements
-    else pack remaining_bits :: packed_field_elements
+  let packed_field_elements, remaining_bits, remaining_length =
+    Array.fold bitstrings ~init:([], [], 0) ~f:(fun (acc, bits, n) bitstring ->
+        let n = n + List.length bitstring in
+        let bits = bits @ bitstring in
+        let acc, bits, n = pack_full_fields acc bits n in
+        (acc, bits, n) )
   in
+  if remaining_length = 0 then packed_field_elements
+  else pack remaining_bits :: packed_field_elements
+
+let pack_to_fields ~size_in_bits ~pack {field_elements; bitstrings} =
+  let max_size = size_in_bits - 1 in
+  let packed_bits = pack_bits ~max_size ~pack {field_elements; bitstrings} in
   Array.append field_elements (Array.of_list_rev packed_bits)
 
 let to_bits ~unpack {field_elements; bitstrings} =
   let field_bits = Array.map ~f:unpack field_elements in
   List.concat @@ Array.to_list @@ Array.append field_bits bitstrings
 
+module Coding = struct
+  (** See https://github.com/CodaProtocol/coda/blob/develop/rfcs/0038-rosetta-construction-api.md for details on schema *)
+
+  (** Serialize a random oracle input with 32byte fields into bytes according to the RFC0038 specification *)
+  let serialize ~string_of_field ~to_bool ~of_bool t =
+    let len_to_string x =
+      String.of_char_list
+        Char.
+          [ of_int_exn @@ ((x lsr 24) land 0xff)
+          ; of_int_exn @@ ((x lsr 16) land 0xff)
+          ; of_int_exn @@ ((x lsr 8) land 0xff)
+          ; of_int_exn @@ (x land 0xff) ]
+    in
+    let len1 = len_to_string @@ Array.length t.field_elements in
+    let fields =
+      (* We only support 32byte fields *)
+      let _ =
+        match t.field_elements with
+        | [|x; _|] ->
+            assert (String.length (string_of_field x) = 32)
+        | _ ->
+            ()
+      in
+      Array.map t.field_elements ~f:string_of_field |> String.concat_array
+    in
+    let len2 =
+      len_to_string
+      @@ Array.sum (module Int) t.bitstrings ~f:(fun x -> List.length x)
+    in
+    let packed =
+      pack_bits t ~max_size:8 ~pack:(fun bs ->
+          let rec go i acc = function
+            | [] ->
+                acc
+            | b :: bs ->
+                go (i + 1) ((acc * 2) + if to_bool b then 1 else 0) bs
+          in
+          let pad =
+            List.init (8 - List.length bs) ~f:(Fn.const (of_bool false))
+          in
+          let combined = bs @ pad in
+          assert (List.length combined = 8) ;
+          go 0 0 combined )
+      |> List.map ~f:Char.of_int_exn
+      |> List.rev |> String.of_char_list
```

Comment:
I'm not 100% sure why I need the List.rev here -- I imagine I actually need it somewhere else. Bonus points for the reviewer if you can explain why this is actually right or where the List.rev should go instead.

Reply:
This is a quirk of `pack_bits`; maybe it should be called `pack_bits_rev`. Note the `Array.of_list_rev packed_bits` above.